rslib/src/storage/tag/mod.rs

```diff
@@ -85,11 +139,49 @@ impl SqliteStorage {
 
     pub(super) fn downgrade_tags_from_schema14(&self) -> Result<()> {
         let alltags = self.all_tags()?;
-        let tagsmap: HashMap<String, Usn> = alltags.into_iter().collect();
+        let tagsmap: HashMap<String, Usn> = alltags.into_iter().map(|t| (t.name, t.usn)).collect();
         self.db.execute(
             "update col set tags=?",
             params![serde_json::to_string(&tagsmap)?],
         )?;
         Ok(())
     }
+
+    pub(super) fn upgrade_tags_to_schema17(&self) -> Result<()> {
+        let tags = self
+            .db
+            .prepare_cached("select tag, usn from tags")?
+            .query_and_then(NO_PARAMS, |r| {
+                Ok(Tag {
+                    name: r.get(0)?,
+                    usn: r.get(1)?,
+                    ..Default::default()
+                })
+            })?
+            .collect::<Result<Vec<Tag>>>()?;
+        self.db.execute_batch(
+            "
```

Comment:
When the sql exceeds a single line, please move it into a separate .sql file, and include_str!() it. I've just added a formatter for .sql files, so we'll get consistent formatting that way, and syntax highlighting in code editors.